src/ray/core_worker/task_event_buffer.h

```diff
@@ -0,0 +1,178 @@
+  /// Constructor
+  ///
+  /// \param gcs_client GCS client
+  /// \param manual_flush Test only flag to disable periodical flushing events.
+  TaskEventBufferImpl(std::unique_ptr<gcs::GcsClient> gcs_client,
+                      bool manual_flush = false);
+
+  void AddTaskStatusEvent(TaskID task_id,
+                          rpc::TaskStatus task_status,
+                          std::unique_ptr<rpc::TaskInfoEntry> task_info,
+                          std::unique_ptr<rpc::TaskStateEntry> task_state_update)
+      LOCKS_EXCLUDED(mutex_) override;
+
+  void AddProfileEvent(TaskID task_id,
+                       rpc::ProfileEventEntry event,
+                       const std::string &component_type,
+                       const std::string &component_id,
+                       const std::string &node_ip_address)
+      LOCKS_EXCLUDED(mutex_) override;
+
+  /// Flush all of the events that have been added since last flush to the GCS.
+  /// If previous flush's gRPC hasn't been replied and `forced` is false, the flush will
+  /// be skipped until the next invocation.
+  ///
+  /// \param forced True if it should be flushed regardless of previous gRPC event's
+  /// state.
+  void FlushEvents(bool forced) LOCKS_EXCLUDED(mutex_) override;
+
+  /// Stop the TaskEventBuffer and it's underlying IO, disconnecting GCS clients.
+  void Stop() LOCKS_EXCLUDED(mutex_) override;
+
+  rpc::TaskEventData GetAllTaskEvents() LOCKS_EXCLUDED(mutex_) override {
+    absl::MutexLock lock(&mutex_);
+    rpc::TaskEventData copy(task_events_data_);
+    return copy;
+  }
+
+ private:
+  /// Mutex guarding task_events_data_.
+  absl::Mutex mutex_;
+
+  /// IO service event loop owned by TaskEventBuffer.
+  instrumented_io_context io_service_;
+
+  /// Work guard to prevent the io_context from exiting when no work.
+  // boost::asio::io_service::work io_work_;
+  boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard_;
+
+  /// Dedicated io thread for running the periodical runner and the GCS client.
+  std::thread io_thread_;
+
+  /// The runner to run function periodically.
+  PeriodicalRunner periodical_runner_;
+
+  /// Client to the GCS used to push profile events to it.
+  std::unique_ptr<gcs::GcsClient> gcs_client_;
+
+  /// Current task event data to be pushed to GCS.
+  rpc::TaskEventData task_events_data_ GUARDED_BY(mutex_);
+
+  /// Flag to toggle event recording on/off.
+  bool recording_on_ = false;
+
+  /// True if there's a pending gRPC call. It's a simple way to prevent overloading
+  /// GCS with too many calls. There is no point sending more events if GCS could not
+  /// process them quick enough.
+  /// TODO(rickyx): When there are so many workers, we might even want to proxy those to
+  /// the agent/raylet to further prevent overloading GCS.
+  std::atomic<bool> grpc_in_progress_ = false;
+
+  /// Stats tracking for debugging and monitoring.
+  size_t total_events_bytes_ = 0;
+  size_t total_num_events_ = 0;
```

Comment:
Guard?

Reply:
those (and others above) are actually accessed from the same thread in current implementation - i will add comment.